feat: exclude declared variable when Object literal completions

chenjigeng opened the pull request:
<!--
Thank you for submitting a pull request!

Please verify that:
* [ ] There is an associated issue in the `Backlog` milestone (**required**)
* [ ] Code is up-to-date with the `master` branch
* [ ] You've successfully run `gulp runtests` locally
* [ ] There are new or updated unit tests validating the change

Refer to CONTRIBUTING.MD for more details.
  https://github.com/Microsoft/TypeScript/blob/master/CONTRIBUTING.md
-->

Fixes #41731

bar9 commented:
test

andrewbranch commented on src/services/completions.ts:
```diff
@@ -1526,6 +1532,44 @@ namespace ts.Completions {
+            }
+
+            let curNode: Node = node.parent;
+            while (curNode && curNode.parent) {
+                if (isPropertyAssignment(curNode) && isObjectLiteralExpression(curNode.parent)) {
+                    curNode = curNode.parent.parent;
+
+                    if (isVariableDeclaration(curNode)) {
+                        return curNode;
+                    }
+                }
+                else {
+                    return undefined;
+                }
+            }
+
+            return undefined;
```
Take a look at `findAncenstor` and see if that lets you write this more concisely.

chenjigeng replied:
But I find that in this case, filterObjectMembersList will not be executed

chenjigeng replied:
> Take a look at `findAncenstor` and see if that lets you write this more concisely.
I used findAncenstor at first, but I found that findAncenstor didn't do much of the work because I had to filter two nodes at a time, one for property assignments and one for object literal expression nodes

andrewbranch reviewed (changes requested):
Thanks for taking a look! Two big things to get started:

1. This needs tests.
2. I would expect zero extra work to be done if we’re not in object literal completions. Right now this is doing some amount of work for every completions request that returns any globals at all. Whatever work is going to be done should probably happen in `filterObjectMembersList`.

chenjigeng commented:
> Thanks for taking a look! Two big things to get started:
> 
> 1. This needs tests.
> 2. I would expect zero extra work to be done if we’re not in object literal completions. Right now this is doing some amount of work for every completions request that returns any globals at all. Whatever work is going to be done should probably happen in `filterObjectMembersList`.

2. But I find that in this case, filterObjectMembersList will not be executed.

andrewbranch commented:
> I find that in this case of the issue, `filterObjectMembersList` will not be executed and I don't know why

Oh, I totally wasn’t thinking. That’s for completions on object literal keys, not values, which could be anything.

The test looks great and answers some of the questions I was going to ask next. I think this looks good overall, just going to leave some suggestions for names and code style.

andrewbranch commented on src/services/completions.ts:
```diff
@@ -1526,6 +1532,44 @@ namespace ts.Completions {
             });
         }
 
+        function findUndeclaredVariableIfExist(node: Node): VariableDeclaration | undefined {
```
```suggestion
        function getVariableDeclarationOfObjectLiteral(property: Node): VariableDeclaration | undefined {
```

andrewbranch commented on src/services/completions.ts:
```diff
@@ -1501,13 +1501,19 @@ namespace ts.Completions {
                     : KeywordCompletionFilters.TypeKeywords;
             }
 
+            const undeclaredVariable = findUndeclaredVariableIfExist(location);
```
```suggestion
            const objectLiteralVariableDeclaration = getVariableDeclarationOfObjectLiteral(location);
```

andrewbranch commented on src/services/completions.ts:
```diff
@@ -1501,13 +1501,19 @@ namespace ts.Completions {
                         return true;
                     }
 
+                    if (isUndeclaredSymbol(undeclaredVariable, symbol)) {
```
```suggestion
                    // Filter out variables from their own initializers, e.g. `const o = { prop: /* no 'o' here */ }`
                    if (objectLiteralVariableDeclaration && contains(symbol.declarations, objectLiteralVariableDeclaration)) {
```

andrewbranch replied:
Actually, in these cases the declaration should always be the valueDeclaration, right? This is a bit more efficient and I think should work:

```suggestion
                    if (objectLiteralVariableDeclaration && symbol.valueDeclaration === objectLiteralVariableDeclaration) {
```

andrewbranch commented on tests/cases/fourslash/completionListInObjectLiteral5.ts:
```diff
@@ -0,0 +1,24 @@
+
+////const o = 'something' 
+////const obj = {
+////    prop: o/*1*/,
```
I just merged #41539—if you merge/rebase with master, you could add a marker for a shorthand property assignment.

chenjigeng replied:
OK~done

andrewbranch reviewed (changes requested):
Now that I’m thinking about this more, it’s really not specific to object literals at all. Variable declaration names should probably never show up in completions for their own initializers unless a function body boundary has been crossed:

```ts
const a = a;
const b = a && b;
const c = [{ prop: [c] }];
```

I suspect there wouldn’t be much performance cost to generalizing to that. You could walk up parent nodes and bail if you hit a `FunctionBody` or an expression that was an `ArrowFunction["body"]`, and return the variable declaration if you hit a `VariableDeclaration["initializer"]`.

chenjigeng commented:
> Now that I’m thinking about this more, it’s really not specific to object literals at all. Variable declaration names should probably never show up in completions for their own initializers unless a function body boundary has been crossed:
> 
> ```ts
> const a = a;
> const b = a && b;
> const c = [{ prop: [c] }];
> ```
> 
> I suspect there wouldn’t be much performance cost to generalizing to that. You could walk up parent nodes and bail if you hit a `FunctionBody` or an expression that was an `ArrowFunction["body"]`, and return the variable declaration if you hit a `VariableDeclaration["initializer"]`.

Yes. It Sounds like good. I can try to do it. But I think this may need many test cases.

chenjigeng commented:
> > Now that I’m thinking about this more, it’s really not specific to object literals at all. Variable declaration names should probably never show up in completions for their own initializers unless a function body boundary has been crossed:
> > ```ts
> > const a = a;
> > const b = a && b;
> > const c = [{ prop: [c] }];
> > ```
> > 
> > 
> > I suspect there wouldn’t be much performance cost to generalizing to that. You could walk up parent nodes and bail if you hit a `FunctionBody` or an expression that was an `ArrowFunction["body"]`, and return the variable declaration if you hit a `VariableDeclaration["initializer"]`.
> 
> Yes. It Sounds like good. I can try to do it. But I think this may need many test cases.

I submit a new PR to do this (based on the current submission). https://github.com/microsoft/TypeScript/pull/42087 I have modified the existing test case and added a new test case. You can help to see if there is any problem

sandersn commented:
I'm going to close this PR since #42087 supersedes it.
